File: backend/src/main/java/com/hst/sitescrapper/interceptor/JwtAuthenticationFilter.java
package com.hst.sitescrapper.interceptor;

import com.hst.sitescrapper.service.JwtTokenProvider;
import com.hst.sitescrapper.service.UserService;
import org.springframework.security.core.Authentication;
import org.springframework.security.core.context.SecurityContextHolder;
import org.springframework.web.filter.GenericFilterBean;

import javax.servlet.FilterChain;
import javax.servlet.ServletException;
import javax.servlet.ServletRequest;
import javax.servlet.ServletResponse;
import javax.servlet.http.HttpServletRequest;
import java.io.IOException;

/**
 * @author dlgusrb0808@gmail.com
 */
public class JwtAuthenticationFilter extends GenericFilterBean {

	private final JwtTokenProvider jwtTokenProvider;
	private final UserService userService;

	public JwtAuthenticationFilter(JwtTokenProvider jwtTokenProvider, UserService userService) {
		this.jwtTokenProvider = jwtTokenProvider;
		this.userService = userService;
	}

	@Override
	public void doFilter(ServletRequest request, ServletResponse response, FilterChain filterChain) throws IOException, ServletException {
		String token = jwtTokenProvider.resolveToken((HttpServletRequest) request);
		if (jwtTokenProvider.validateToken(token)) {
			Authentication authentication = userService.getAuthentication(token);
			SecurityContextHolder.getContext().setAuthentication(authentication);
		}
		filterChain.doFilter(request, response);
	}
}

File: backend/src/main/kotlin/com/hst/scrapper/global/filter/JwtAuthenticationFilter.kt
package com.hst.scrapper.global.filter

import com.hst.scrapper.modules.user.application.UserService
import com.hst.scrapper.modules.user.domain.repo.AuthTokenRepository
import org.slf4j.LoggerFactory
import org.springframework.security.authentication.UsernamePasswordAuthenticationToken
import org.springframework.security.core.context.SecurityContextHolder
import org.springframework.security.web.authentication.WebAuthenticationDetailsSource
import org.springframework.web.filter.OncePerRequestFilter
import javax.servlet.FilterChain
import javax.servlet.http.HttpServletRequest
import javax.servlet.http.HttpServletResponse

/**
 * @author dlgusrb0808@gamil.com
 */
class JwtAuthenticationFilter(
    val authTokenRepository: AuthTokenRepository,
    val userService: UserService
) : OncePerRequestFilter() {

    companion object {
        val log = LoggerFactory.getLogger(JwtAuthenticationFilter::class.java)
    }

    override fun doFilterInternal(
        request: HttpServletRequest,
        response: HttpServletResponse,
        filterChain: FilterChain
    ) {
        log.info("Check authentication for {}", request.requestURI)
        val token = authTokenRepository.getToken(request)
        if (authTokenRepository.isValidToken(token)) {
            val userId = authTokenRepository.getUserIdFromToken(token!!)
            val userDetail = userService.getUser(userId)
            val authentication = UsernamePasswordAuthenticationToken(userDetail, userDetail.password, userDetail.authorities)
            authentication.details = WebAuthenticationDetailsSource().buildDetails(request)
            SecurityContextHolder.getContext().authentication = authentication
        }
        filterChain.doFilter(request, response)
    }

}